volume/store/store.go

```diff
@@ -67,22 +67,42 @@ func (s *VolumeStore) List() ([]volume.Volume, []string, error) {
 		name := normaliseVolumeName(v.Name())
 
 		s.locks.Lock(name)
-		storedV, exists := s.getNamed(name)
-		if !exists {
+		// make sure that this volume still exists after grabbing a lock on its name
+		// this is to prevent a race where you can delete a volume between when it's
+		// listed above and cached below because the lock on its name is not held
+		if !s.existsInDriver(v) {
+			s.locks.Unlock(name)
+			continue
+		}
+		storedV, existsInCache := s.getNamed(name)
+		if !existsInCache {
 			s.setNamed(v, "")
```

Comment:
Actually, Get also uses the cache through getVolume. Since it seems to work, we should probably not remove it. I guess I can just remove the caching during List as you suggested.